### PR TITLE
Add Typed-binding-pattern support for incremental parser

### LINKED ISSUE #22938: Add typed-binding-pattern support to foreach statement; incremental parser
Add typed-binding-pattern support to foreach statement; incremental parser

<code>typed-binding-pattern := inferable-type-descriptor binding-pattern</code>
<code>inferable-type-descriptor := type-descriptor | var</code>

Currently foreach statement only has <code>foreach typedesc varName in action-or-expr block-stmt</code>

### LINKED ISSUE #23079: Add capture-binding-pattern support for typed-binding-pattern
Add capture-binding-pattern support for typed-binding-pattern

<code>capture-binding-pattern := variable-name</code>
<code>variable-name := identifier</code>

### LINKED ISSUE #23081: Add list-binding-pattern for typed-binding-pattern
Add list-binding-pattern for typed-binding-pattern

<code>list-binding-pattern := [ list-member-binding-patterns ]</code>
<code>list-member-binding-patterns :=
   binding-pattern (, binding-pattern)* [, rest-binding-pattern]
   | [ rest-binding-pattern ]</code>

### LINKED ISSUE #23083: Add rest-binding-pattern support for typed-binding-pattern
Add rest-binding-pattern support for typed-binding-pattern

<code>rest-binding-pattern := ... variable-name</code>

### PR DESCRIPTION
**Pull request by m36dot**
Add typed-binding-pattern support for foreach statement
Add capture-binding-pattern support for typed-binding-pattern
Add list-binding-pattern support for typed-binding-pattern
Add rest-binding-pattern support for typed-binding-pattern

Closes #22938
Closes #23079
Closes #23083 
Closes #23081 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -1983,7 +1991,10 @@ private STNode parseComplexTypeDescriptor(STNode typeDesc, ParserRuleContext con
                 return parseComplexTypeDescriptor(parseOptionalTypeDescriptor(typeDesc), context);
             // If next token after a type descriptor is <code>[</code> then it is an array type descriptor
             case OPEN_BRACKET_TOKEN:
-                return parseComplexTypeDescriptor(parseArrayTypeDescriptor(typeDesc), context);
+                if (isListBindingPattern()) { // return typed binding attern if we find a list-b-p
+                    return STNodeFactory.createTypedBindingPatternNode(typeDesc, parseListBindingPattern());
+                }
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1996]: It is not correct to create typed-binding-patterns here. Reason is, this is a generic method to parse type-descs. 

But typed-binding-patterns are not supported for all places where a type is supported.

e.g: type-cast, func-return, type-parameter, etc. there are many places where a type-desc can appear without a binding pattern.

Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1996]: hmmm, that makes sense.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 2]
```diff
@@ -8423,4 +8434,337 @@ private STNode parseParenthesisedTypeDesc() {
...
+            switch (child.kind) {
+                case DECIMAL_INTEGER_LITERAL:
+                case HEX_INTEGER_LITERAL:
+                case ASTERISK_TOKEN:
+                case CAPTURE_BINDING_PATTERN:
+                    return false;
+                default: // cases like [ and {
+                    return true;
+            }
+        } else {
```
[the remark this answers is not in the document]
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:8751]: the only time we have to solve this problem is to decide between array type desc and list-b-p in typed binding patterns; if there are no typed binding patterns then it proceeds like usual(like it did before our code).

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 3]
```diff
@@ -9343,4 +9375,340 @@ private STNode parseIntersectionTypeDescriptor(STNode leftTypeDesc, ParserRuleCo
...
+
+        if (isListBindingPatternDefinitively(arrayDescOrListBindingPattern)) { // ambiguous means T[a]
+            return STNodeFactory.createTypedBindingPatternNode(typeDesc,
+                    cleanListBindingPattern(arrayDescOrListBindingPattern));
+        }
+
+        if (isFollowTypedBindingPattern(peek().kind)) {
+            return STNodeFactory.createTypedBindingPatternNode(typeDesc,
+                    cleanListBindingPattern(arrayDescOrListBindingPattern));
+        }
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9623]: body of this if block is same as above. can we merge the two?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 4]
```diff
@@ -9343,4 +9375,340 @@ private STNode parseIntersectionTypeDescriptor(STNode leftTypeDesc, ParserRuleCo
...
+            }
+            bindingPatterns.add(listBindingPatternRhs);
+            listBindingPatternMember = parselistBindingPatternMember();
+            bindingPatterns.add(listBindingPatternMember);
+            token = peek();
+        }
+        STNode closeBracket = parseCloseBracket();
+
+        //seperating out the restbindingpattern
+        STNode restBindingPattern = STNodeFactory.createEmptyNode();
```
[the remark this answers is not in the document]
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9486]: ok

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 5]
```diff
@@ -9343,4 +9375,340 @@ private STNode parseIntersectionTypeDescriptor(STNode leftTypeDesc, ParserRuleCo
...
+    }
+
+    private STNode parseListBindingpatternRhs(SyntaxKind nextTokenKind) {
+        switch (nextTokenKind) {
+            case COMMA_TOKEN:
+                return parseComma();
+            case CLOSE_BRACKET_TOKEN:
+                return null;
+            default:
+                Solution solution = recover(peek(), ParserRuleContext.TYPE_DESC_IN_TUPLE_RHS);
```
[the remark this answers is not in the document]
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9510]: sorry didnt notice

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 3 above)
Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9623]: Can you add a comment mentioning what are we checking here, and why we need that?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 6]
```diff
@@ -9343,4 +9375,340 @@ private STNode parseIntersectionTypeDescriptor(STNode leftTypeDesc, ParserRuleCo
...
+        typeDesc = mergeTypeDescAndListBindingPattern(typeDesc, arrayDescOrListBindingPattern);
+        if (peek().kind == SyntaxKind.OPEN_BRACKET_TOKEN) {
+            //this means arrayDescOrBLP is a array type desc
+            return parseArrayTypeDescOrListBindingPattern(typeDesc);
+        } else {
+            return STNodeFactory.createTypedBindingPatternNode(typeDesc,
+                    parseBindingPattern());
+        }
+
+
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9634]: Extra new lines

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 7]
```diff
@@ -9343,4 +9375,342 @@ private STNode parseIntersectionTypeDescriptor(STNode leftTypeDesc, ParserRuleCo
...
+        }
+
+        if (numberOfChildren == 1) {
+            STNode child = childArray.childInBucket(0);
+            switch (child.kind) {
+                case LIST_BINDING_PATTERN:
+                case REST_BINDING_PATTERN:
+                    return true;
+                default: // cases like [ and {
+                    return false;
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9695]: I think its safe to do the reverse of this. 
i.e: if the only-child is captured-binding-patterns, then return true.
That way we don't have to update this method, when we add new binding patterns.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 8]
```diff
@@ -9343,4 +9375,342 @@ private STNode parseIntersectionTypeDescriptor(STNode leftTypeDesc, ParserRuleCo
...
+            switch (child.kind) {
+                case DECIMAL_INTEGER_LITERAL:
+                case HEX_INTEGER_LITERAL:
+                case ASTERISK_TOKEN:
+                    this.errorHandler.reportInvalidNode(null,
+                            "invalid Node " + child.kind + " in list-binding-pattern");
+                    cleanedChildren.add(STNodeFactory.createMissingToken(SyntaxKind.IDENTIFIER_TOKEN));
+                    break;
+                default:
+                    cleanedChildren.add(child);
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9668]: Regardless whether its a valid child or an invalid child, we need to keep it. That is required to regenerate the source-code using the syntax tree.

So we can simply rename this method as "validateListBindingPattern", and only log an error for invalid children, but let them stay.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 7 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9695]: sure

File: compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json [hunk 9]
```diff
@@ -2848,6 +2844,78 @@
...
             ]
+        },
+        {
+            "name": "TypedBindingPatternNode",
+            "base": "Node",
+            "kind": "TYPED_BINDING_PATTERN",
+            "attributes": [
+                {
+                    "name": "typeDescriptor",
+                    "type": "Node"
```
Comment (SupunS) [inline compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json:2855]: Should be TypeDescriptorNode

File: compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json [hunk 10]
```diff
@@ -2848,6 +2844,78 @@
...
+            "isAbstract": true
+        },
+        {
+            "name": "CaptureBindingPatternNode",
+            "base": "BindingPatternNode",
+            "kind": "CAPTURE_BINDING_PATTERN",
+            "attributes": [
+                {
+                    "name": "variableName",
+                    "type": "Token",
```
Comment (SupunS) [inline compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json:2875]: SimpleNameReferenceNode

File: compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json [hunk 11]
```diff
@@ -2848,6 +2844,78 @@
...
+            "base": "Node",
+            "kind": "REST_BINDING_PATTERN",
+            "attributes": [
+                {
+                    "name": "ellipsisToken",
+                    "type": "Token"
+                },
+                {
+                    "name": "variableName",
+                    "type": "Token"
```
Comment (SupunS) [inline compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json:2916]: SimpleNameReferenceNode

File: compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json [hunk 12]
```diff
@@ -2848,6 +2844,78 @@
...
+        },
+        {
+            "name": "CaptureBindingPatternNode",
+            "base": "BindingPatternNode",
+            "kind": "CAPTURE_BINDING_PATTERN",
+            "attributes": [
+                {
+                    "name": "variableName",
+                    "type": "SimpleNameReferenceNode",
+                    "isOptional": true
```
Comment (SupunS) [inline compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json:2876]: Is this optional?

**Pull request merged**